### ISSUE TITLE
FIRESTORE INTERNAL ASSERTION FAILED: enterCheckedOperation may not be called when an operation is in progress

### ISSUE THREAD

**Issue by businessengine** [firebase/firebase-ios-sdk#988]
### [READ] Step 1: Are you in the right place?

  * For issues or feature requests related to __the code in this repository__
    file a Github issue.
    * If this is a __feature request__ make sure the issue title starts with "FR:".
  * For general technical questions, post a question on [StackOverflow](http://stackoverflow.com/)
    with the firebase tag.
  * For general Firebase discussion, use the [firebase-talk](https://groups.google.com/forum/#!forum/firebase-talk)
    google group.
  * For help troubleshooting your application that does not fall under one
    of the above categories, reach out to the personalized
    [Firebase support channel](https://firebase.google.com/support/).

### [REQUIRED] Step 2: Describe your environment

  * Xcode version: 9.2
  * Firebase SDK version: 4.11.0
  * Firebase Component: Firestore (Auth, Core, Database, Firestore, Messaging, Storage, etc)
  * Component version: 0.10.4

#### Steps to reproduce:

What happened? How can we make the problem occur?
```
*** Assertion failure in -[FSTDispatchQueue enterCheckedOperation:], third_party/firebase/ios/Source/Firestore/Source/Util/FSTDispatchQueue.mm:192
2018-03-28 10:03:30.389690+0700 BusinessEngine[16302:12429566] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'FIRESTORE INTERNAL ASSERTION FAILED: enterCheckedOperation may not be called when an operation is in progress'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010e5f612b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x000000010d0daf41 objc_exception_throw + 48
	2   CoreFoundation                      0x000000010e5fb2f2 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x000000010cb7bd69 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 193
	4   BusinessEngine                      0x0000000103b83460 -[FSTDispatchQueue enterCheckedOperation:] + 270
	5   libdispatch.dylib                   0x000000010f4c32f7 _dispatch_call_block_and_release + 12
	6   libdispatch.dylib                   0x000000010f4c433d _dispatch_client_callout + 8
	7   libdispatch.dylib                   0x000000010f4ca754 _dispatch_continuation_pop + 967
	8   libdispatch.dylib                   0x000000010f4c8b85 _dispatch_async_redirect_invoke + 780
	9   libdispatch.dylib                   0x000000010f4d0102 _dispatch_root_queue_drain + 772
	10  libdispatch.dylib                   0x000000010f4cfda0 _dispatch_worker_thread3 + 132
	11  libsystem_pthread.dylib             0x000000010f98f1ca _pthread_wqthread + 1387
	12  libsystem_pthread.dylib             0x000000010f98ec4d start_wqthread + 13
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by wilhuff** [comment]
Can you tell us some more about how and when this happens? This stack trace doesn't really tell us anything because the assertion being made here should hold with a stack that's in this configuration. More context will help us understand.

**Comment by businessengine** [comment]
Hi @wilhuff, 
Full stack (pic 01)
![image](https://user-images.githubusercontent.com/578995/38021827-2008d1a0-32a8-11e8-85f9-4223ad4eaef0.png)
Pic 02
![image](https://user-images.githubusercontent.com/578995/38021860-2e0bd658-32a8-11e8-806f-9c7428756fec.png)
Pic 04:
![image](https://user-images.githubusercontent.com/578995/38021898-41ed8a04-32a8-11e8-9aa9-1bfc75b4ae48.png)
Thanks Gil.

**Comment by businessengine** [comment]
Reverted back to `'Firebase/Firestore', '4.10.1'`, no issues.

**Comment by wilhuff** [comment]
Hi, I've been trying to reproduce this issue and haven't had any luck. The stack shows the standard flow to set up a connection wherein Firestore requests the token for the currently signed in user and then passes it on to gRPC to start up the connection. I'm able to sign in with a variety of accounts and Firestore is able to read data as expected.

Could you please give some more details:
  * What are you doing that triggers this issue?
  * Is there a sequence of steps required?
  * Does it happen every time?

**Comment by olejnjak** [comment]
Hi, I've got the same problem. 

My setup looks the same as @businessengine. But I am setting custom dispatch queue for Firebase callbacks using this code in my AppDelegate

```swift
func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
    FirebaseApp.configure()
        
    let bgQueue = DispatchQueue.global(qos: .background)
        
    let settings = FirestoreSettings()
    settings.dispatchQueue = bgQueue
    Firestore.firestore().settings = settings
        
    Storage.storage().callbackQueue = bgQueue
        
    window = UIWindow(frame: UIScreen.main.bounds)
    window?.makeKeyAndVisible()
    appFlow.start()
        
    return true
}
```

With this setup my app crashes on _FIRESTORE INTERNAL ASSERTION FAILED: enterCheckedOperation may not be called when an operation is in progress_ assert as mentioned in this issue.

If I stick to the main thread it works fine.

Is there any more information I can provide? Or is there anything I'm doing wrong?

**Comment by morganchen12** [comment]
@mikelehen could this be caused by the new `enterCheckedOperation` changes if an `FSTDispatchQueue` is created with a concurrent queue like in the above example?

**Comment by olejnjak** [comment]
I'll create a minimal example project if it would help (?) or maybe I could provide my project which is affected by this.

**Comment by mikelehen** [comment]
@morganchen12 Almost certainly, but we're not sure how.

@olejnjak We looked into this before but weren't able to reproduce it or come up with a likely cause, so a minimal example would be extremely helpful.  Thanks!

**Comment by olejnjak** [comment]
Here it is. It is not exactly minimal example as I am now busy. Just packed my project which contains almost nothing so it can be considered minimal.

[CarManager.zip](https://github.com/firebase/firebase-ios-sdk/files/1887534/CarManager.zip)

**Comment by olejnjak** [comment]
I was trying with the latest release (4.12) but I am unable to install it using Carthage...
```
*** Downloading FirebaseStorageBinary.framework binary at "4.12.0"
A shell task (/usr/bin/env unzip -uo -qq -d /var/folders/83/2fpf32vn49n1pc6y8bbyv8c80000gn/T/carthage-archive.aNDxjo /Users/olejnjak/Library/Caches/org.carthage.CarthageKit/binaries/FirebaseFirestoreBinary/4.12.0/Firestore-87a804ab561d91db.zip) failed with exit code 9:
[/Users/olejnjak/Library/Caches/org.carthage.CarthageKit/binaries/FirebaseFirestoreBinary/4.12.0/Firestore-87a804ab561d91db.zip]
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of /Users/olejnjak/Library/Caches/org.carthage.CarthageKit/binaries/FirebaseFirestoreBinary/4.12.0/Firestore-87a804ab561d91db.zip or
        /Users/olejnjak/Library/Caches/org.carthage.CarthageKit/binaries/FirebaseFirestoreBinary/4.12.0/Firestore-87a804ab561d91db.zip.zip, and cannot find /Users/olejnjak/Library/Caches/org.carthage.CarthageKit/binaries/FirebaseFirestoreBinary/4.12.0/Firestore-87a804ab561d91db.zip.ZIP, period.
```

Deleting `~/Library/Caches/org.carthage.CarthageKit` did not help

**Comment by paulb777** [comment]
@olejnjak That really looks like a stale cache error. Are you able to download https://dl.google.com/dl/firebase/ios/carthage/4.12.0/Firestore-87a804ab561d91db.zip from a browser?

**Comment by olejnjak** [comment]
But the problem persists

**Comment by paulb777** [comment]
I'm not sure why the cache isn't updating.  One workaround might be manually updating the zero lengthed ./FirebaseFirestoreBinary/4.12.0/Firestore-87a804ab561d91db.zip in the Cache with the one downloaded from a web url.

**Comment by olejnjak** [comment]
By the problem I mean the "FIRESTORE INTERNAL ASSERTION FAILED: enterCheckedOperation may not be called when an operation is in progress" exception :-)

**Comment by var-const** [comment]
@olejnjak Just wanted to let you know that I could reproduce the issue using the project you uploaded. I [fixed](https://github.com/firebase/firebase-ios-sdk/commit/9f02fa6b2ee1bfac85b20907693c0afdb1d274b5#diff-8470e7048f44b5ce9fcfbb0024967f09) a race condition which could be connected to the error, but haven't verified yet that it resolves the error you're seeing. I'll continue working on this issue and will follow up here.

**Comment by olejnjak** [comment]
Cool, so do we agree that it's an error inside Firestore and not the client code?

**Comment by wilhuff** [comment]
Based on our current line of investigation, yes that seems to be the case, but we haven't yet confirmed one way or the other.

**Comment by var-const** [comment]
@olejnjak To give a progress update, we've been able to identify the issue. It's triggered by configuring `FirestoreSettings` to use a concurrent queue, so for a quick workaround, you could change that to use a serial queue (which is the default). I'll continue working on fixing this and will post an update when I have it.

**Comment by olejnjak** [comment]
Nice, thanks :-)

**Comment by businessengine** [comment]
Cool, we will use the default queue in the meantime.

**Comment by var-const** [comment]
Should be fixed as of release 5.1.0.
